File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    int rec(int n,vector<int> & dp)
    {
        if(n==1||n==2||n==0)
        return dp[n]=n;

        if(dp[n]!=-1)
        return dp[n];

        return dp[n]=rec(n-1,dp)+rec(n-2,dp);
        //once we come to these stairs only one way is there
    }

    int climbStairs(int n) {
        vector<int> dp(n+1,-1);
        // dp[0]=0;
        // dp[1]=1;
        // dp[2]=2;
        //what if n=1 dp[2] won't exist
        return rec(n,dp);
    }
};

File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    public int rec(int n,List<Integer> dp)
    {
        if(dp.get(n)!=-1)
        return dp.get(n);

        int result=rec(n-1,dp)+rec(n-2,dp);
        dp.set(n,result);
        return result;

    }

    public int climbStairs(int n) {
        List<Integer> dp = new ArrayList<>(Collections.nCopies(n + 1, -1));

        if (n == 0) return 0;
        if (n == 1) return 1;
        if (n == 2) return 2;

        dp.set(0, 0);
        dp.set(1, 1);
        dp.set(2, 2);
        return rec(n,dp);
    }
}